### PR TITLE
Item assignment in csc/csr

### LINKED ISSUE #3115: 'csc_matrix' object does not support item assignment
I am using sparse matrices using CuPy. If I define the csc_matrix using scipy, I can assign the matrix after its definition. But, I am not able to do it with CuPy. Below is my code:

```
data = np.ones((3,4))
diags = np.array([-1,0,1])
M= spdiags(data, diags, 4, 4)
M = sp.csc_matrix(M)
print(M[0,0])
M[0,0] = 5
print(M[0,0])
```
I get 1.0, 5.0 as the output. If I try to do same thing in CuPy I get an error.
```
import cupy as cp
import cupyx as cpx
datac = cp.ones((3,4))
diagsc = cp.array(diags)
Mc = cpx.scipy.sparse.spdiags(datac, diagsc, 4, 4)
Mc = cpx.scipy.sparse.csc_matrix(Mc)

Mc = cpx.scipy.sparse.csc_matrix(Mc)
print(Mc.get()[0,0])
Mc[0,0] = 5
print(Mc[0,0])
```
I get 1.0 as th eoutput and then the error: 'csc_matrix' object does not support item assignment.

Am I doing something wrong?

Also, the assignment operation with csc_matrix is expensive in SciPy. But, I can use lil_matrix in scipy which is much faster. Does CuPy have a sparse lil_matirx option?

Thank you.

### PR DESCRIPTION
**Pull request by Dahlia-Chehata**
closes #3115

### CODE REVIEW

**Comment by Dahlia-Chehata** [comment]
Tests will be added later

**Comment by emcastillo** [comment]
Please, test set items using ellipsis and slices.

Thanks!

File: cupyx/scipy/sparse/compressed.py [hunk 1]
```diff
@@ -33,6 +33,51 @@ class _compressed_sparse_matrix(sparse_data._data_matrix,
...
+    _compress_setitem_complex_kern = core.ElementwiseKernel(
+        'T value_real, T value_imag, S ind, int32 index',
+        'raw T real, raw T imag',
+        '''
+          atomicExch(&real[index], value_real);
+          atomicExch(&imag[index], value_imag);
+        ''',
+        'compress_setitem_complex',
+        preamble='''
+            #if __CUDA_ARCH__ < 600
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py]: This code is the same as above, store it in a `setitem_preamble` variable.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py [hunk 2]
```diff
@@ -1367,3 +1366,52 @@ def test_getitem_slice_stop_too_small(self, xp, sp):
...
+    def test_setitem_int_int(self, xp, sp):
+        matrix = _make(xp, sp, self.dtype)
+        matrix[0, 0] = 10
+        assert matrix[0, 0] == 10
+
+    @testing.numpy_cupy_equal(sp_name='sp')
+    def test_setitem_complex(self, xp, sp):
+        matrix = _make_complex(xp, sp, self.dtype)
+        matrix[1, 2] = 10 + 2j
+        assert matrix[0, 0] == 10 + 2j
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: This test won't pass

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csr.py [hunk 3]
```diff
@@ -1493,3 +1493,75 @@ def test_getitem_slice_stop_too_small(self, xp, sp):
...
+    def test_setitem_int_int(self, xp, sp):
+        matrix = _make(xp, sp, self.dtype)
+        matrix[0, 0] = 10
+        assert matrix[0, 0] == 10
+
+    @testing.numpy_cupy_equal(sp_name='sp')
+    def test_setitem_complex(self, xp, sp):
+        matrix = _make_complex(xp, sp, self.dtype)
+        matrix[1, 2] = 10 + 2j
+        assert matrix[0, 0] == 10 + 2j
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_csr.py]: Ditto

**Comment by emcastillo** [comment]
You have essentially copied all the logic from `get_item`. But I am afraid this doesn't work in this case.

Setting a value in a sparse matrix requires to alter all the data structures holding metadata and the actual data. In this PR you don't allow to add new items to the matrix, which is not correct for this feature.
 
I am afraid that for example, if you wan't to set an element which is currently 0 in a csr sparse matrix this code won't work.

This is how a csr matrix looks like:
image

https://matteding.github.io/2019/04/25/sparse-matrices/

File: cupyx/scipy/sparse/compressed.py (hunk 1 above)
Comment (emcastillo) [reply cupyx/scipy/sparse/compressed.py]: Also add a comment with the reference where you get the code from
`https://www.micc.unifi.it/bertini/download/gpu-programming-basics/2017/gpu_cuda_5.pdf`
(I don't know if this is the original source ...)

**Comment by Dahlia-Chehata** [comment]
@emcastillo, So if an item is already present in the sparse matrix, and it was set to zero, it should remain present as in the sparse matrix which is the original behavior in SciPy, right?

**Comment by Dahlia-Chehata** [comment]
I am thinking of converting to dense matrix, setting the required values and then  back to the sparse structure. If this is a good approach, I will start modifying it

**Comment by emcastillo** [comment]
no, you have to modify in place the data structures and eliminate the entries that are 0ed.

Please, read the link that I posted above and refer to the scikit source code for implementation details.

**Comment by Dahlia-Chehata** [comment]
> no, you have to modify in place the data structures and eliminate the entries that are 0ed.
>  
> Please, read the link that I posted above and refer to the scikit source code for implementation details.

Ok, I will read it and check the source code. But when I tried to set a value to 0 ( previously not present in the sparse matrix or had a non zero value before) in SciPy. The item was present in the sparse structure when printing the matrix and with a zero value. That is why I asked

**Comment by emcastillo** [comment]
>>  The item was present in the sparse structure when printing the matrix and with a zero value.

Yes, although this works it is not consistent with the definition of a compressed sparse matrix where only non-zero entries are stored 😃

File: cupy/manipulation/add_remove.py [hunk 4]
```diff
@@ -1,13 +1,21 @@
...
+from cupy import core
 
 # TODO(okuta): Implement delete
 
 
 # TODO(okuta): Implement insert
 
-
-# TODO(okuta): Implement append
+def append(tup, values, axis=None):
```
Comment (emcastillo) [inline cupy/manipulation/add_remove.py]: This function is implemented in https://github.com/cupy/cupy/pull/1802/files#diff-4fd534cefc7099c784eefd47425e197bR13

Please don't add unrelated features in the same PR.

File: cupyx/scipy/sparse/compressed.py [hunk 5]
```diff
@@ -634,38 +687,111 @@ def __setitem__(self, slices, value, cs_type):
...
+        minor = cupy.array(minor, dtype=self.indices.dtype,
+                           copy=False, ndmin=1).ravel()
+        value = cupy.array(value, dtype=self.dtype,
+                           copy=False, ndmin=1).ravel()
+        # check boundaries
+        major_max = cupy.amax(major)
+        major_min = cupy.amin(major)
+        minor_max = cupy.amax(minor)
+        minor_min = cupy.amin(minor)
+        if major_max >= major_size or major_min < - major_size \
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py]: This is causing unnecessary device synchronization which can greatly hurt performance.
Since you have the actual major and minor and shape objects in the cpu main memory, you can use them for comparisons.

File: cupyx/scipy/sparse/compressed.py [hunk 6]
```diff
@@ -52,31 +51,85 @@ class _compressed_sparse_matrix(sparse_data._data_matrix,
...
+                }
             }
-            #endif
-            ''')
+            outptr[tid] = offset;
+            ret = 0;
+     }
+    ''', 'data_offsets')
+
+    loaded_from_source = r'''
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py]: use a more descriptive name such as `_lower_bound_kernel`

**Comment by emcastillo** [comment]
I would be grateful if you can explain the logic behind each function. Some comments in the code would be helpful

File: cupyx/scipy/sparse/compressed.py (hunk 5 above)
Comment (Dahlia-Chehata) [reply cupyx/scipy/sparse/compressed.py]: Actually major and minor here are arrays of rows and columns indices used in list assignment

File: cupy/manipulation/add_remove.py (hunk 4 above)
Comment (Dahlia-Chehata) [reply cupy/manipulation/add_remove.py]: Sure, I need it in the insertion part that is why I put it here. I will replace it with the function provided in the link if it is not merged yet

**Comment by Dahlia-Chehata** [comment]
Thank you for reviewing. I will put detailed comments for each function soon

File: cupyx/scipy/sparse/compressed.py (hunk 5 above)
Comment (emcastillo) [reply cupyx/scipy/sparse/compressed.py]: You can calculate them in the python objects,
doing if comparisons of cupy values force device synchronization and it can be pretty expensive.

File: cupy/cuda/memory_hooks/line_profile.py [hunk 7]
```diff
@@ -25,7 +25,8 @@ class LineProfileHook(memory_hook.MemoryHook):
 
             _root (4.00KB, 4.00KB)
               lib/python3.6/unittest/__main__.py:18:<module> (4.00KB, 4.00KB)
-                lib/python3.6/unittest/main.py:255:runTests (4.00KB, 4.00KB)
+                lib/python3.6/unittest/__init__.py:255:
```
Comment (emcastillo) [inline cupy/cuda/memory_hooks/line_profile.py]: why this change?

File: cupyx/scipy/sparse/compressed.py [hunk 8]
```diff
@@ -33,6 +33,92 @@ class _compressed_sparse_matrix(sparse_data._data_matrix,
         ''',
         'compress_getitem_complex')
 
+    # https: // www.micc.unifi.it / bertini / download
+    # / gpu - programming - basics / 2017 / gpu_cuda_5.pdf
+    setitem_preamble = '''
+            #if __CUDA_ARCH__ < 600
+            __device__ double atomicExch(double* address, double val)
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py]: I think this function can be greatly simplified by just using reinterpret_cast

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py [hunk 9]
```diff
@@ -1,7 +1,6 @@
 import unittest
-
-import numpy
 import pytest
+import numpy
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: Revert import order

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py (hunk 2 above)
Comment (emcastillo) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: How about this?

File: cupyx/scipy/sparse/compressed.py [hunk 10]
```diff
@@ -558,13 +644,253 @@ def _get_major_slice(self, major):
...
+            row, col = slices[0], slice(None)
+        else:
+            raise IndexError('invalid number of indices')
+
+        major, minor = self._swap(row, col)
+        major_size, minor_size = self._swap(*self._shape)
+
+        if type(major) is slice or type(minor) is slice:
+            raise ValueError("Unsupported assignment")
+        else:
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py]: you can remove the else: here, after raise this code will never be executed

File: cupyx/scipy/sparse/compressed.py [hunk 11]
```diff
@@ -558,13 +644,253 @@ def _get_major_slice(self, major):
...
+        # if found, else -1
+        offsets = cupy.empty(values.size, dtype=self.indices.dtype)
+        repeat = self.sample_offsets(major_size, minor_size,
+                                     values.size, major, minor, offsets)
+        # In case of duplicates entries, eliminate entries by summation
+        # and repeat the operation
+        if repeat == 1:
+            self.sum_duplicates()
+            self.sample_offsets(major_size, minor_size,
+                                values.size, major, minor, offsets)
```
Comment (emcastillo) [inline cupyx/scipy/sparse/compressed.py]: Why is this necessary?

File: cupyx/scipy/sparse/compressed.py [hunk 12]
```diff
@@ -33,6 +33,92 @@ class _compressed_sparse_matrix(sparse_data._data_matrix,
...
 
+    # https: // www.micc.unifi.it / bertini / download
+    # / gpu - programming - basics / 2017 / gpu_cuda_5.pdf
+    setitem_preamble = '''
+            #if __CUDA_ARCH__ < 600
+            __device__ double atomicExch(double* address, double val)
+            {
+                unsigned long long int* address_as_ull =
+                                          (unsigned long long int*)address;
+                unsigned long long int old = *address_as_ull, compare;
```
Comment (leofang) [inline cupyx/scipy/sparse/compressed.py]: why isn't `compare` declared?

Comment (leofang) [inline cupyx/scipy/sparse/compressed.py]: Could you split the declaration of `compare` into two lines? This is a very confusing statement.

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py (hunk 2 above)
Comment (Dahlia-Chehata) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: Why it won't pass ?

Comment (emcastillo) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: You are setting item [1,2] but checking [0,0]

Comment (Dahlia-Chehata) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: Sorry, haven't noticed that

File: cupyx/scipy/sparse/compressed.py (hunk 12 above)
Comment (emcastillo) [reply cupyx/scipy/sparse/compressed.py]: I think that `uint64_t address_as_ull = reinterpret_cast<uint64_t>(address);` should be enough

File: tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py (hunk 2 above)
Comment (emcastillo) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: It is strange that the test passed with this in your environment

Comment (Dahlia-Chehata) [reply tests/cupyx_tests/scipy_tests/sparse_tests/test_csc.py]: yes, it only showed some warning but not an error. I am debugging the cause of this now.

**Comment by Dahlia-Chehata** [comment]
Performance comparison of list assignment
`SciPy:  __setitem__         :  1420.748 us   +/-138.612 (min: 1269.375 / max: 9060.028) us   1427.665 us   +/-140.917 (min: 1275.232 / max: 9068.160) us
CuPy: __setitem__         :  1379.985 us   +/-70.124 (min: 1269.334 / max: 4080.980) us   1386.917 us   +/-74.607 (min: 1275.648 / max: 4088.480) us
`

File: cupyx/scipy/sparse/compressed.py (hunk 11 above)
Comment (Dahlia-Chehata) [reply cupyx/scipy/sparse/compressed.py]: So do you suggest to remove it and set duplicate entries each time when found ?

File: cupyx/scipy/sparse/compressed.py (hunk 12 above)
Comment (Dahlia-Chehata) [reply cupyx/scipy/sparse/compressed.py]: > Could you split the declaration of `compare` into two lines? This is a very confusing statement.

Do you mean to re-declare the data type ?

**Comment by emcastillo** [comment]
Have you tried with several sizes?
I would like to see the performance results for say, 1000, 10000, 100000 and 500000 elements.
Thanks!
If the size is small, we expect performance to be worse than numpy/scipy.

**Comment by Dahlia-Chehata** [comment]
These results were for n=10000 for the following example
matrix [[1, 2, 3, 0, 1, 3], [2, 1, 2, 0, 0, 1]] = [5, 0, 12, 5, 2, 9]
I will try larger values of n

**Comment by Dahlia-Chehata** [comment]
> **n =1000**
SciPy:  __setitem__       :  1053.414 us   +/-157.928 (min: 1014.712 / max: 3983.991) us   1058.139 us   +/-158.049 (min: 1019.168 / max: 3990.272) us
CuPy: __setitem__         :  1040.992 us   +/-90.520 (min: 1010.135 / max: 3690.532) us   1045.812 us   +/-90.618 (min: 1015.680 / max: 3696.224) us

**n = 5000**
SciPy:  __setitem__       :  1052.410 us   +/-115.087 (min: 1020.867 / max: 3751.660) us   1057.462 us   +/-115.517 (min:  999.488 / max: 3757.472) us
CuPy: __setitem__         :  1048.592 us   +/-109.850 (min: 1014.130 / max: 3764.885) us   1053.616 us   +/-109.935 (min: 1018.912 / max: 3770.560) us

**n = 10000**
SciPy:  __setitem__       :  1062.869 us   +/-136.334 (min: 1027.299 / max: 7871.480) us   1067.911 us   +/-136.602 (min: 1005.056 / max: 7879.136) us
CuPy: __setitem__         :  1052.396 us   +/-29.638 (min: 1021.682 / max: 1458.768) us   1057.441 us   +/-29.786 (min: 1026.944 / max: 1464.736) us

**n = 100000**
SciPy:  __setitem__       :  1052.969 us   +/-56.578 (min: 1021.416 / max: 8685.867) us   1058.082 us   +/-57.332 (min: 1026.304 / max: 8692.704) us
CuPy: __setitem__         :  1045.694 us   +/-34.475 (min: 1015.872 / max: 3858.949) us   1050.750 us   +/-34.621 (min: 1020.736 / max: 3864.544) us

**n = 500000**
SciPy:  __setitem__       :  1036.899 us   +/-48.362 (min: 1006.364 / max: 4467.046) us   1041.730 us   +/-49.385 (min:  988.064 / max: 4477.088) us
CuPy: __setitem__         :  1032.975 us   +/-47.087 (min: 1003.562 / max: 4653.503) us   1037.761 us   +/-47.090 (min:  988.512 / max: 4664.576) us

File: cupyx/scipy/sparse/compressed.py (hunk 11 above)
Comment (Dahlia-Chehata) [reply cupyx/scipy/sparse/compressed.py]: I eliminated that part

**Comment by emcastillo** [comment]
changing just the `n` of the matrix has no impact because you are increasing the number of zeros and keeping the non-zeros constant. So the execution times are the same because entries that are zero are not tracked at all.

You need to increase the number of non-zero elements.

**Comment by Dahlia-Chehata** [comment]
The following results were done using n = 10 where n is the number of repetitions and the list was filled with random numbers for each list size
**list size = 1000**
SciPy:  __setitem__         : 62223.847 us   +/-19413.174 (min: 4757.487 / max:72004.195) us  62232.039 us   +/-19413.532 (min: 4764.704 / max:72012.321) us
CuPy: __setitem__         : 51045.143 us   +/-23570.895 (min: 4792.529 / max:70292.275) us  51052.746 us   +/-23571.159 (min: 4799.424 / max:70300.705) us
**list size = 10000**
SciPy:  __setitem__         : 511959.374 us   +/-255478.783 (min: 9492.486 / max:733896.758) us  511971.843 us   +/-255480.639 (min: 9500.448 / max:733914.062) us
CuPy: __setitem__         : 416344.135 us   +/-265726.922 (min: 9216.744 / max:601814.716) us  416354.763 us   +/-265728.972 (min: 9225.184 / max:601826.172) us
**list size = 100000**
SciPy:  __setitem__         : 1721775.022 us   +/-2554768.482 (min:44191.947 / max:5632601.273) us  1721796.336 us   +/-2554778.823 (min:44203.743 / max:5632638.672) us
CuPy: __setitem__         : 1154870.733 us   +/-2216361.341 (min:44489.589 / max:5593468.612) us  1154887.357 us   +/-2216371.668 (min:44500.065 / max:5593505.371) us
**list size = 1000000**
SciPy:  __setitem__         : 6032337.818 us   +/-16879022.505 (min:402792.835 / max:56669404.973) us  6032381.689 us   +/-16879112.473 (min:402806.458 / max:56669718.750) us
CuPy: __setitem__         : 410040.921 us   +/-7411.889 (min:403919.516 / max:430660.061) us  410054.736 us   +/-7411.815 (min:403932.800 / max:430673.645) us

**Comment by Dahlia-Chehata** [comment]
Is this testing sufficient or should I change the matrix size itself ?

**Comment by emcastillo** [comment]
Thanks! the test seems good to me
I assume you are measuring with `cupyx.time.repeat`?

Will you also try the zeroing & addition approach? If possible I would like to see both.
Thanks again!

**Comment by Dahlia-Chehata** [comment]
**list size = 1000**

SciPy:  __setitem__         :  8338.613 us   +/-377.754 (min: 8033.822 / max: 9428.667) us   8345.085 us   +/-378.058 (min: 8039.680 / max: 9436.000) us
CuPy: __setitem__         :  9383.629 us   +/-2262.886 (min: 7980.722 / max:13908.296) us   9390.755 us   +/-2264.364 (min: 7986.464 / max:13920.896) us

**list size = 10000**

SciPy:  __setitem__         : 11699.934 us   +/-729.362 (min:11145.791 / max:13837.378) us  11707.789 us   +/-728.772 (min:11152.000 / max:13844.288) us
CuPy: __setitem__         : 11518.263 us   +/-964.094 (min:11100.711 / max:14401.475) us  11524.589 us   +/-964.488 (min:11107.072 / max:14408.960) us

**list size = 100000**

SciPy:  __setitem__         : 45128.798 us   +/-848.671 (min:44658.848 / max:47623.215) us  45135.478 us   +/-848.729 (min:44664.928 / max:47630.047) us
CuPy: __setitem__         : 44807.189 us   +/-466.978 (min:44338.600 / max:46121.830) us  44813.876 us   +/-467.223 (min:44345.089 / max:46129.440) us

**list size = 1000000**

SciPy:  __setitem__         : 396749.410 us   +/-6981.837 (min:389082.897 / max:412930.872) us  396759.552 us   +/-6982.248 (min:389092.468 / max:412940.948) us
CuPy: __setitem__         : 392472.052 us   +/-2934.542 (min:389969.189 / max:400127.765) us  392481.604 us   +/-2934.858 (min:389978.546 / max:400138.367) us

**Comment by emcastillo** [comment]
It is strange that execution times are the very same than scipy 😂

**Comment by Dahlia-Chehata** [comment]
> It is strange that execution times are the very same than scipy joy

I know this seems pretty suspicious :joy: but these are the true results I've got. Here is another run of the script:
**1000**
SciPy:  __setitem__         :  8158.059 us   +/-100.895 (min: 8059.939 / max: 8435.645) us   8164.477 us   +/-100.931 (min: 8065.888 / max: 8441.728) us
CuPy: __setitem__         :  8138.873 us   +/-168.085 (min: 7923.455 / max: 8459.837) us   8145.002 us   +/-168.308 (min: 7929.312 / max: 8466.528) us
**10000**
SciPy:  __setitem__         : 11235.268 us   +/-118.512 (min:11017.472 / max:11451.098) us  11241.654 us   +/-118.563 (min:11023.648 / max:11457.536) us
CuPy: __setitem__         : 11179.567 us   +/-51.101 (min:11079.481 / max:11254.308) us  11185.869 us   +/-51.084 (min:11085.440 / max:11260.480) us
**100000**
SciPy:  __setitem__         : 45559.527 us   +/-870.327 (min:44925.754 / max:47962.334) us  45566.451 us   +/-870.341 (min:44932.575 / max:47969.311) us
CuPy: __setitem__         : 45387.318 us   +/-199.024 (min:45008.236 / max:45659.469) us  45394.141 us   +/-199.202 (min:45014.561 / max:45666.527) us
**1000000**
SciPy:  __setitem__         : 402015.587 us   +/-2677.346 (min:396314.763 / max:405914.776) us  402033.496 us   +/-2677.420 (min:396332.581 / max:405932.190) us
CuPy: __setitem__         : 400166.018 us   +/-5331.489 (min:393540.280 / max:409780.540) us  400183.087 us   +/-5330.793 (min:393558.228 / max:409796.570) us

If you find something strange or causing unnecessary overhead in the implementation of the addition and zeroing approach, I can modify it

**Comment by emcastillo** [comment]
Sorry for the dela, the main source of performance degradation here is the COO->CSR conversions, It would be faster if we could create the matrix directly in CSR format.

**Comment by kmaehashi** [comment]
Closing, will be done in another PR #3486
